File: spring-data-requery-java/src/main/java/com/coupang/springframework/data/requery/repository/RequeryRepository.java
package com.coupang.springframework.data.requery.repository;

import com.coupang.springframework.data.requery.core.RequeryOperations;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Sort;
import org.springframework.data.repository.NoRepositoryBean;
import org.springframework.data.repository.PagingAndSortingRepository;
import org.springframework.data.repository.query.QueryByExampleExecutor;

import java.util.List;

/**
 * RequeryRepository
 *
 * @author debop@coupang.com
 * @since 18. 6. 4
 */
@NoRepositoryBean
public interface RequeryRepository<T, ID>
    extends PagingAndSortingRepository<T, ID>, QueryByExampleExecutor<T>, RequeryWhereExecutor<T> {

    @Autowired
    RequeryOperations getOperations();

    @Override
    List<T> findAll();

    @Override
    List<T> findAll(Sort sort);

    @Override
    List<T> findAllById(Iterable<ID> ids);

    @Override
    <S extends T> List<S> saveAll(Iterable<S> entities);

    <S extends T> S upsert(S entity);

    <S extends T> List<S> upsertAll(Iterable<S> entities);

    void deleteInBatch(Iterable<T> entities);

    int deleteAllInBatch();

    T getOne(ID id);

    @Override
    <S extends T> List<S> findAll(Example<S> example);

    @Override
    <S extends T> List<S> findAll(Example<S> example, Sort sort);

}

File: src/main/kotlin/com/coupang/springframework/data/requery/repository/RequeryRepository.kt
package com.coupang.springframework.data.requery.repository

import org.springframework.data.domain.Example
import org.springframework.data.domain.Sort
import org.springframework.data.repository.NoRepositoryBean
import org.springframework.data.repository.PagingAndSortingRepository
import org.springframework.data.repository.query.QueryByExampleExecutor

/**
 * RequeryRepository
 *
 * @author debop@coupang.com
 * @since 18. 5. 29
 */
@NoRepositoryBean
interface RequeryRepository<T, ID>: PagingAndSortingRepository<T, ID>, QueryByExampleExecutor<T> {

    override fun findAll(): Iterable<T>

    override fun findAll(sort: Sort): Iterable<T>

    override fun findAllById(ids: Iterable<ID>): Iterable<T>

    override fun <S: T> saveAll(entities: Iterable<S>): Iterable<S>

    fun <S: T> upsert(entity: S): S

    fun <S: T> upsertAll(entities: Iterable<S>): Iterable<S>

    fun deleteInBatch(entities: Iterable<T>)

    fun deleteAllInBatch()

    fun getOne(id: ID): T

    override fun <S: T> findAll(example: Example<S>): Iterable<S>

    override fun <S: T> findAll(example: Example<S>, sort: Sort): Iterable<S>
}